spark/src/main/scala/org/apache/spark/sql/delta/commands/merge/ClassicMergeExecutor.scala

```diff
@@ -256,15 +319,23 @@ trait ClassicMergeExecutor extends MergeOutputGeneration {
     // with value `true`, one to each side of the join. Whether this field is null or not after
     // the outer join, will allow us to identify whether the joined row was a
     // matched inner result or an unmatched result with null on one side.
-    val joinedDF = {
+    val joinedBaseDF = {
       var sourceDF = getMergeSource.df
       if (deduplicateCDFDeletes.enabled && deduplicateCDFDeletes.includesInserts) {
         // Add row index for the source rows to identify inserted rows during the cdf deleted rows
         // deduplication. See [[deduplicateCDFDeletes()]]
         sourceDF = sourceDF.withColumn(SOURCE_ROW_INDEX_COL, monotonically_increasing_id())
       }
-      val left = sourceDF
-        .withColumn(SOURCE_ROW_PRESENT_COL, Column(incrSourceRowCountExpr))
+      val left = if (writeUnmodifiedRows) {
+        sourceDF.withColumn(SOURCE_ROW_PRESENT_COL, Column(incrSourceRowCountExpr))
+      } else {
+        sourceDF
+          .withColumn(SOURCE_ROW_PRESENT_COL, Column(incrSourceRowCountExpr))
+          // In some cases, the optimiser (incorrectly) decides to omit the metrics column.
+          // This causes issues in the source determinism validation. We work around the issue by
+          // adding a redundant dummy filter to make sure the column is not pruned.
+          .filter(SOURCE_ROW_PRESENT_COL)
```

Comment:
Can we always add the dummy filter and not differentiate between the two cases? I imagine an always true filter isn't going to be noticeable here.